src/cpp/rtps/writer/StatefulWriter.cpp

```diff
@@ -1000,14 +1105,56 @@ void StatefulWriter::process_acknack(const GUID_t reader_guid, uint32_t ack_coun
                 }
                 else if(SequenceNumber_t{0, 0} == remote_reader->get_low_mark() && sn_set.isSetEmpty() && !final_flag)
                 {
-                    send_heartbeat_to_nts(*remote_reader, true);
+                    send_heartbeat_to_nts(*remote_reader);
                 }
 
-                // Check if all CacheChange are acknowledge, because a user could be waiting
+                // Check if all CacheChange are acknowledged, because a user could be waiting
                 // for this, of if VOLATILE should be removed CacheChanges
                 check_acked_status();
             }
             break;
         }
     }
 }
+
+void StatefulWriter::lifespan_expired()
+{
+    std::unique_lock<std::recursive_mutex> lock(*mp_mutex);
+
+    CacheChange_t* earliest_change;
+    if (!mp_history->get_earliest_change(&earliest_change))
+    {
+        return;
+    }
+
+    auto source_timestamp = system_clock::time_point() + nanoseconds(earliest_change->sourceTimestamp.to_ns());
+    auto now = system_clock::now();
+
+    // Check that the earliest change in the history has actually expired as
+    // the change that started the timer could have been removed from the history
+    if (now - source_timestamp < keep_duration_us_)
+    {
+        auto interval = source_timestamp - now + keep_duration_us_;
+        lifespan_timer_->update_interval_millisec((double)duration_cast<milliseconds>(interval).count());
+        lifespan_timer_->restart_timer();
+        return;
+    }
+
+    // The earliest sample has expired
+    mp_history->remove_change_g(earliest_change);
```

Comment:
No always it should be removed. It has to change to ACKNOWLEDGE, and depending of the rest of QoS it should be removed or not.

Reply:
OK, I've replaced that line with:
```
for(const auto& remote_reader : matched_readers)
{
    remote_reader->acked_changes_set(earliest_change->sequenceNumber);
}
```

What will happen if after the `StatefulWriter` has marked the change as acknowledged it receives an ack from the reader saying that it is missing that change? Will everything work fine?

Also, because I shouldn't be removing samples from the history, the blackbox tests that I had added do not make sense anymore. I'll remove them as I can't think of a good way to test this, unless there is an easy way to simulate the case where the reader does not get the samples?